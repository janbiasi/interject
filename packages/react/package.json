{
	"name": "@interject/react",
	"version": "0.0.1",
	"description": "Use the interject framework together with React",
	"main": "./dist/interject-react.js",
	"types": "./dist/meta/src/index.d.ts",
	"scripts": {
		"test-watch": "jest --watch",
		"test": "jest --verbose",
		"test-ci": "jest --verbose --ci --clearCache",
		"build": "run-s build-*",
		"build-source": "webpack",
		"build-declarations": "tsc --project tsconfig.meta.json"
	},
	"repository": {
		"type": "git",
		"url": "git+https://github.com/janbiasi/flawless-di.git"
	},
	"keywords": ["di", "dependency", "injection", "dependency", "injection", "inject", "react", "react di"],
	"author": "Jan Biasi <biasijan@gmail.com> (http://janbiasi.ch)",
	"license": "MIT",
	"bugs": {
		"url": "https://github.com/janbiasi/flawless-di/issues"
	},
	"devDependencies": {
		"reflect-metadata": "0.1.12",
		"@interject/config-typescript": "*",
		"@interject/config-webpack": "*",
		"@interject/core": "*",
		"@types/jest": "23.3.3",
		"@types/react": "16.4.16",
		"jest": "23.6.0",
		"npm-run-all": "4.1.3",
		"react": "16.5.2",
		"react-dom": "16.5.2",
		"ts-jest": "23.10.4",
		"tslint": "5.11.0",
		"typescript": "3.1.1",
		"webpack": "4.20.2",
		"webpack-cli": "3.1.2",
		"enzyme": "3.7.0",
		"enzyme-adapter-react-16": "1.6.0"
	},
	"peerDependencies": {
		"@interject/core": ">=0.1.0",
		"react": ">=16.0.0<17.0.0",
		"react-dom": ">=16.0.0<17.0.0"
	},
	"jest": {
		"roots": ["<rootDir>", "<rootDir>/test"],
		"transform": {
			"^.+\\.tsx?$": "ts-jest"
		},
		"setupFiles": ["./setupTests.ts"],
		"collectCoverage": true,
		"testRegex": "(/__tests__/.*|(\\.|/)(test|spec))\\.tsx?$",
		"collectCoverageFrom": ["**/*.ts"],
		"coveragePathIgnorePatterns": ["<rootDir>/test/__snapshots__", "<rootDir>/src/index.ts", "<rootDir>/dist/meta"],
		"moduleFileExtensions": ["js", "jsx", "ts", "tsx", "json"],
		"globals": {
			"ts-jest": {
				"diagnostics": {
					"ignoreCodes": [151001]
				}
			}
		}
	}
}
