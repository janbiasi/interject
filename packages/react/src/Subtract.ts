/**!
 * @see https://github.com/piotrwitek/utility-types/blob/dec0bd052ed18db8cdf70eb9ed5cb6e90a8111f2/src/mapped-types.ts
 */

export type SetDifference<A, B> = A extends B ? never : A;

export type SetComplement<A, A1 extends A> = SetDifference<A, A1>;

export type Subtract<T extends T1, T1 extends object> = Pick<T, SetComplement<keyof T, keyof T1>>;
