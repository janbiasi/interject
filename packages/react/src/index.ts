export { withInterject } from './withInterject';
export {Â ReactProvidableModule } from './utils';