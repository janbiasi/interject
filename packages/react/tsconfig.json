{
	"extends": "./node_modules/@interject/config-typescript/tsconfig.json",
	"compilerOptions": {
		"rootDir": "./",
		"baseUrl": "./",
		"declaration": true,
		"jsx": "react",
		"experimentalDecorators": true
	},
	"include": ["**/*.ts", "**/*.tsx"],
	"exclude": ["dist"]
}
