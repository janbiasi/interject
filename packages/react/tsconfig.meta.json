{
	"compilerOptions": {
		"rootDir": "./",
		"baseUrl": "./",
		"outDir": "./dist/meta",
		"declaration": true,
		"emitDeclarationOnly": true,
		"resolveJsonModule": true,
		"lib": ["es2015"],
		"target": "es5",
		"jsx": "react",
		"experimentalDecorators": true
	},
	"include": ["./src/**/*.ts"],
	"exclude": ["dist"]
}
