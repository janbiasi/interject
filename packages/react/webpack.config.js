// @ts-check
const createWebpackConfig = require('@interject/config-webpack');

module.exports = createWebpackConfig({
	path: __dirname,
	name: 'interject-react',
	library: 'InterjectReactPlugin',
});
