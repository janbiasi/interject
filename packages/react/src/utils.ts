import { IConstructable } from "@interject/core/dist/meta/src/Constructable";

export type ReactProvidableModule = {
    provideAs?: string
}

export const getAnchorProp = (Factory: IConstructable<any>): string => {
    if (typeof (<ReactProvidableModule>Factory).provideAs === 'string') {
        return (<ReactProvidableModule>Factory).provideAs;
    }

    return Factory.toString();
}