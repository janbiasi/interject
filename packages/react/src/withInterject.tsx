import * as React from 'react';
import { Subtract } from './Subtract';
import { internalBootstrapFactory } from '@interject/core';
import { getAnchorProp } from './utils';

export type InterjectModule<T extends any> = { new (...args: any[]): T };

export interface IWithInterjectProps {
    [moduleProvideToken: string]: any;
}

export interface IWithInterjectState {}

export interface IInterjectComponentPersistance {
    providers: any[];
}

/** @internal */
const modulePersistanceMap: Map<InterjectModule<any>, IInterjectComponentPersistance> = new Map();

export function withInterject<T extends InterjectModule<any>, P extends IWithInterjectProps>(Component: React.ComponentType<P>, ModuleFactory?: T) {
    return class InterjectionComponent extends React.Component<Subtract<P, IWithInterjectProps>, IWithInterjectState> {
        state: IWithInterjectState = {};

        aggreagateModuleInstance(ModuleFactory?: InterjectModule<any>) {
            if (!ModuleFactory) {
                return null;
            }

            const cached = modulePersistanceMap.get(ModuleFactory);
            if (cached) {
                return cached;
            }

            const bootstrappedModule = internalBootstrapFactory(ModuleFactory);
            modulePersistanceMap.set(ModuleFactory, bootstrappedModule);
            return bootstrappedModule;
        }

        render() {
            if (!ModuleFactory) {
                return <Component {...this.props}/>;
            }

            const inst = this.aggreagateModuleInstance(ModuleFactory);
            const key = getAnchorProp(inst);
            const injectedModuleProps = {
                [key]: inst
            };
            
            return <Component {...this.props} {...injectedModuleProps} />;
            
        }
    }
}