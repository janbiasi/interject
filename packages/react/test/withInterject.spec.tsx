import * as React from 'react';
import { shallow, render } from 'enzyme';
import { Module, InjectionToken } from '@interject/core';
import { withInterject } from '../src/withInterject';

const TEST_TOKEN = new InjectionToken<string>('a test token');

describe('withInterject', () => {
	it('should render a react component', () => {
		const Stub = () => <p>react-stub-noop</p>;
		const EnhancedStub = withInterject(Stub);

		expect(shallow(<Stub />)).toMatchSnapshot();
		expect(shallow(<EnhancedStub />)).toMatchSnapshot();
	});

	it('should inject an empty module into a react component', () => {
		@Module()
		class MyModule {
			public static provideAs = 'MyModule';
		}

		const Component = () => <p>MyComponent</p>;
		const MyComponent = withInterject(Component, MyModule);

		expect(shallow(<MyComponent />)).toMatchSnapshot();
	});

	it('should inject a module with providers into a react component', () => {
		@Module({
			providers: [{ provide: TEST_TOKEN, useValue: 'test' }],
		})
		class MyModule {
			public provideAs = 'MyModule';
			public value: string;

			constructor(value: string) {
				this.value = value;
			}
		}

		const Component = ({ MyModule }: { MyModule: MyModule }) => {
			return <p>Value is {MyModule.value}</p>;
		};
		const MyComponent = withInterject(Component, MyModule);

		expect(shallow(<MyComponent />)).toMatchSnapshot();
		expect(render(<MyComponent />)).toMatchSnapshot();
	});

	it('should inject a module as singleton into a react component using persistant state', () => {
		@Module({
			providers: [{ provide: TEST_TOKEN, useValue: 'test' }],
		})
		class MyModule {
			static creationCount: number = 0;
			public provideAs = 'MyModule';
			public value: string;

			constructor(value: string) {
				MyModule.creationCount++;
				this.value = value;
			}
		}

		const ComponentA = ({ MyModule }: { MyModule: MyModule }) => <p>A: {MyModule.value}</p>;
		const ComponentB = ({ MyModule }: { MyModule: MyModule }) => <p>A: {MyModule.value}</p>;
		const A = withInterject(ComponentA, MyModule);
		const B = withInterject(ComponentB, MyModule);

		expect(shallow(<A />)).toMatchSnapshot();
		expect(shallow(<B />)).toMatchSnapshot();
		expect(render(<A />)).toMatchSnapshot();
		expect(render(<B />)).toMatchSnapshot();
		expect(MyModule.creationCount).toEqual(1);
	});
});
