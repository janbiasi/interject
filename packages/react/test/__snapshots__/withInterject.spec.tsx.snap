// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withInterject should inject a module as singleton into a react component using persistant state 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InterjectionComponent />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "MyModule": MyModule {
        "provideAs": "MyModule",
        "value": "test",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "MyModule": MyModule {
          "provideAs": "MyModule",
          "value": "test",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`withInterject should inject a module as singleton into a react component using persistant state 2`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InterjectionComponent />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "MyModule": MyModule {
        "provideAs": "MyModule",
        "value": "test",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "MyModule": MyModule {
          "provideAs": "MyModule",
          "value": "test",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`withInterject should inject a module as singleton into a react component using persistant state 3`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "data": "A: test",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "p",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`withInterject should inject a module as singleton into a react component using persistant state 4`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "data": "A: test",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "p",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`withInterject should inject a module with providers into a react component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InterjectionComponent />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "MyModule": MyModule {
        "provideAs": "MyModule",
        "value": "test",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "MyModule": MyModule {
          "provideAs": "MyModule",
          "value": "test",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`withInterject should inject a module with providers into a react component 2`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "data": "Value is test",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "p",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`withInterject should inject an empty module into a react component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InterjectionComponent />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "MyModule": MyModule {
        "provideAs": "MyModule",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "MyModule": MyModule {
          "provideAs": "MyModule",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`withInterject should render a react component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Stub />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "react-stub-noop",
    },
    "ref": null,
    "rendered": "react-stub-noop",
    "type": "p",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "react-stub-noop",
      },
      "ref": null,
      "rendered": "react-stub-noop",
      "type": "p",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`withInterject should render a react component 2`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InterjectionComponent />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
